File: myblog-server/src/main/java/com/site/blog/context/ConfigContextHolder.java
package com.site.blog.context;

import cn.hutool.core.convert.Convert;
import cn.hutool.core.util.ObjectUtil;
import cn.hutool.core.util.StrUtil;
import com.site.blog.exception.ServiceException;
import lombok.extern.slf4j.Slf4j;

/**
 * @author yanni
 * @date time 2022/6/17 21:07
 * @modified By:
 */
@Slf4j
public class ConfigContextHolder {

    public static String websiteName() {
        return getSysConfigWithDefault("websiteName", String.class, "我的博客");
    }

    public static String domain() {
        return getSysConfigWithDefault("domain", String.class, "http://localhost:2801");
    }
    public static <T> T getSysConfigWithDefault(String configCode, Class<T> clazz, T defaultValue) {
        String configValue = ConfigContext.me().getStr(configCode);
        if (ObjectUtil.isEmpty(configValue)) {
            // 将默认值加入到缓存常量
            log.warn(">>> 系统配置blog_config表中存在空项，configCode为：{}，系统采用默认值：{}", configCode, defaultValue);
            ConfigContext.me().put(configCode, defaultValue);
            return defaultValue;
        } else {
            try {
                return Convert.convert(clazz, configValue);
            } catch (Exception e) {
                return defaultValue;
            }
        }
    }
    public static <T> T getSysConfig(String configCode, Class<T> clazz, Boolean nullThrowExp) {
        log.info(ConfigContext.me().toString());

        String configValue = ConfigContext.me().getStr(configCode);
        if (ObjectUtil.isEmpty(configValue)) {
            if (Boolean.TRUE.equals(nullThrowExp)) {
                String format = StrUtil.format(">>> getSysConfig 系统配置blog_config表中存在空项，configCode为：{}", configCode);
                log.error(format);
                throw new ServiceException(404, format);
            } else {
                return null;
            }
        } else {
            try {
                return Convert.convert(clazz, configValue);
            } catch (Exception e) {
                if (Boolean.TRUE.equals(nullThrowExp)) {
                    String format = StrUtil.format(">>> 系统配置blog_config表中存在格式错误的值，configCode={}，configValue={}", configCode, configValue);
                    log.error(format);
                    throw new ServiceException(404, format);
                } else {
                    return null;
                }
            }
        }
    }
}

File: myblog-kt/src/main/kotlin/com/site/blog/context/ConfigContextHolder.kt
package com.site.blog.context

import cn.hutool.core.convert.Convert
import cn.hutool.core.util.ObjectUtil
import cn.hutool.core.util.StrUtil
import com.site.blog.context.ConfigContext.me
import com.site.blog.exception.ServiceException
import org.slf4j.LoggerFactory

/**
 * @author yanni
 * @date time 2022/6/17 21:07
 * @modified By:
 */

object ConfigContextHolder {
    val log = LoggerFactory.getLogger(this.javaClass)
    @JvmStatic
    fun websiteName(): String {
        return getSysConfigWithDefault("websiteName", String::class.java, "我的博客")
    }

    @JvmStatic
    fun domain(): String {
        return getSysConfigWithDefault("domain", String::class.java, "http://localhost:2801")
    }

    fun <T> getSysConfigWithDefault(configCode: String?, clazz: Class<T>?, defaultValue: T): T {
        val configValue = me().getStr(configCode)
        return if (ObjectUtil.isEmpty(configValue)) {
            // 将默认值加入到缓存常量
            ConfigContextHolder.log.warn(
                ">>> 系统配置blog_config表中存在空项，configCode为：{}，系统采用默认值：{}",
                configCode,
                defaultValue
            )
            me().put(configCode, defaultValue)
            defaultValue
        } else {
            try {
                Convert.convert(clazz, configValue)
            } catch (e: Exception) {
                defaultValue
            }
        }
    }

    fun <T> getSysConfig(configCode: String?, clazz: Class<T>?, nullThrowExp: Boolean): T? {
        ConfigContextHolder.log.info(me().toString())
        val configValue = me().getStr(configCode)
        return if (ObjectUtil.isEmpty(configValue)) {
            if (java.lang.Boolean.TRUE == nullThrowExp) {
                val format =
                    StrUtil.format(">>> getSysConfig 系统配置blog_config表中存在空项，configCode为：{}", configCode)
                ConfigContextHolder.log.error(format)
                throw ServiceException(404, format)
            } else {
                null
            }
        } else {
            try {
                Convert.convert(clazz, configValue)
            } catch (e: Exception) {
                if (java.lang.Boolean.TRUE == nullThrowExp) {
                    val format = StrUtil.format(
                        ">>> 系统配置blog_config表中存在格式错误的值，configCode={}，configValue={}",
                        configCode,
                        configValue
                    )
                    ConfigContextHolder.log.error(format)
                    throw ServiceException(404, format)
                } else {
                    null
                }
            }
        }
    }
}